### PR TITLE
We Can Go Deeper

### PR DESCRIPTION
**Pull request by CodaFi**
This pull request simplifies the JSON machinery and adds the ability to subscript deeper into nested structures using the new `JSONKeypath` Monoid.

### CODE REVIEW

File: Swiftz/Operators.swift [hunk 1]
```diff
@@ -130,6 +138,16 @@ infix operator <+> {
 	associativity right
 }
 
+/// MARK: Data.JSON
+
+/// Retrieve | Retrieves a value from a dictionary of JSON values using a given keypath.
+///
+/// If the given keypath is not present or the retrieved value is not of the appropriate type, this
+/// function returns `.None`.
+infix operator <! {
```
Comment (CodaFi) [inline Swiftz/Operators.swift]: I'm open to ideas on this one.  Aeson uses `.:`, which Swift throws a fit at.  In fact, any permutation of dots and colons is a no go.  Maybe `<?` to emphasize this is the total version of retrieve and use `<!` as a version that forces the Optional?

File: SwiftzTests/JSONSpec.swift [hunk 2]
```diff
@@ -11,30 +11,66 @@ import Swiftz
...
+	let x : Double
+	let y : Double 
+	
+	static func create(x : Double) -> Double -> Coordinate {
+		return { y in Coordinate(x: x, y: y) }
+	}
+	
+	static func fromJSON(x: JSONValue) -> Coordinate? {
+		return Coordinate.create	<^> x <! "map" <> "coordinates" <> "xPart" 
+									<*> x <! "map" <> "coordinates" <> "yPart"
```
Comment (CodaFi) [inline SwiftzTests/JSONSpec.swift]: *sigh*, 8 space tabs again...

File: SwiftzTests/UserExample.swift [hunk 3]
```diff
@@ -10,45 +10,37 @@ import Swiftz
...
-			let r = d["attrs"]  >>- JDictionary<String, JString>.fromJSON
-			// alternatively, if n && a && t... { return User(n!, a!, ...
-			return (User.create <^> n <*> a <*> t <*> r)
-		default:
-			return .None
-		}
+		return User.create	<^> x <! "name" 
+							<*> x <! "age"
+							<*> x <! "tweets" 
+							<*> x <! "attrs" <> "one" // A nested keypath
```
Comment (CodaFi) [inline SwiftzTests/UserExample.swift]: See *sigh*ing above.

File: Swiftz/JSON.swift [hunk 4]
```diff
@@ -357,9 +373,31 @@ public struct JDictionary<A, B: JSON where B.J == A> : JSON {
...
+
+private func resolveKeypath(lhs : Dictionary<String, JSONValue>, rhs : JSONKeypath) -> JSONValue? {
+	if rhs.path.isEmpty {
+		return .None
+	}
+	
+	if let o = lhs[rhs.path.first!] {
+		switch o {
+		case let .JSONObject(d) where rhs.path.count > 1:
+			let tail = Array(rhs.path[1..<rhs.path.count])
```
Comment (CodaFi) [inline Swiftz/JSON.swift]: Probably should make this a function in `ArrayExt`

File: Swiftz/Operators.swift (hunk 1 above)
Comment (pthariensflame) [inline Swiftz/Operators.swift]: That sounds good to me!

**Comment by CodaFi** [comment]
I hope this makes it easier that much easier to knock out #58

Ready for review.

**Comment by pthariensflame** [comment]
This will take a while.  :P

**Comment by pthariensflame** [comment]
This looks fine to me, now!  I wouldn't be surprised if there were spelling errors somewhere, though; I couldn't hold the whole thing in my head at once as I usually can with pull requests.  :P

**Comment by CodaFi** [comment]
Welp, now I'm paranoid.  I'll deal with typos before the next point release.  There's still more I want to do with this before 1.4.

**Comment by CodaFi** [comment]
Nevertheless, :boat:

**Pull request merged**